### PR TITLE
Fix: ToB-21 (unsafe casts)

### PR DESCRIPTION
**Pull request by ChiTimesChi**
**Description**
- Fixed ToB-21 by adding safe casts where the overflow could potentially happen. 
- Documented the usage of unsafe casts where the overflow could not happen due to any of the invariants.
- Also see #1429 for the variables that might need their type adjusted later on to avoid possible reverts on overflows.

## Summary by CodeRabbit

Refactor:
- Introduced a new library `ChainContext` to centralize the retrieval of block timestamp and chain ID, improving consistency and modularity.
- Replaced `uint256` with `uint8` for `stateIndex` parameters across multiple contracts and functions, optimizing memory usage.

New Features:
- Added `ChainContext` library that provides functions for accessing chain context variables.

Bug Fixes:
- Fixed potential overflow issues by using the `SafeCast` library for safe type casting.

Tests:
- Added new test cases to check for overflow conditions in chain ID and block timestamp.

Chores:
- Removed unused import statements and added necessary ones for new libraries.

Note: These changes are primarily code-level optimizations and bug fixes, and they are unlikely to have a direct visible impact on end-users.

### CODE REVIEW

**Review by trajan0x — CHANGES_REQUESTED** [review]

File: packages/contracts-core/test/harnesses/libs/ChainContextHarness.t.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+contract ChainContextHarness {
+    function blockNumber() public view returns (uint40) {
+        return ChainContext.blockNumber();
+    }
+
+    function blockTimestamp() public view returns (uint40) {
+        return ChainContext.blockTimestamp();
+    }
+
+    function chainId() public view returns (uint32) {
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply packages/contracts-core/test/harnesses/libs/ChainContextHarness.t.sol:15]: Oops, this was on harness. Ignore

File: packages/contracts-core/contracts/libs/ChainContext.sol [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/libs/ChainContext.sol:2]: Yes, let's suppress this. 
- As a side note, opened this issue: #1471

File: packages/contracts-core/contracts/libs/stack/GasData.sol [hunk 3]
```diff
@@ -74,6 +74,7 @@ library GasDataLib {
         Number etherPrice_,
         Number markup_
     ) internal pure returns (GasData) {
+        // Number type wraps uint16, so could safely be casted to uint96
         // forgefmt: disable-next-item
         return GasData.wrap(
             uint96(Number.unwrap(gasPrice_)) << SHIFT_GAS_PRICE |
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/libs/stack/GasData.sol:80]: These constants are literally defined on the same page (40 lines above), so no extra explanation here is necessary

File: packages/contracts-core/contracts/manager/BondingManager.sol [hunk 4]
```diff
@@ -88,8 +91,8 @@ contract BondingManager is AgentManager, InterfaceBondingManager {
         if (status.flag == AgentFlag.Unknown) {
             // Unknown address could be added to any domain
             // New agent will need to be added to `_agents` list: could not have more than 2**32 agents
-            if (_agents.length >= type(uint32).max) revert MerkleTreeFull();
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/manager/BondingManager.sol:91]: It's not removed, but rather the revert message is changed (now the generic message from `SafeCast` is used)

[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/manager/BondingManager.sol:91]: In practice I don't think it's possible to have more than `2**32` historical agents, though it's a theorecically possible over-time DoS attack, just need to determine how much time it will take for the attacker to constantly stake-unstake using a fresh address every time (and taking one agent index at a time).

- This is tracked in #1429

File: packages/contracts-core/contracts/manager/BondingManager.sol [hunk 5]
```diff
@@ -88,8 +91,8 @@ contract BondingManager is AgentManager, InterfaceBondingManager {
...
-            index = uint32(_agents.length);
+            // TODO: consider using more than 32 bits for agent indexes
+            index = _agents.length.toUint32();
             // Current leaf for index is bytes32(0), which is already assigned to `leaf`
             _agents.push(agent);
             _domainAgents[domain].push(agent);
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/manager/BondingManager.sol:98]: `.toUint32()` is a function from `SafeCast` library, which handles the reverts, so this check would be a duplicate.

**Comment by ChiTimesChi** [comment]
The failing test in 17b3169 is actually implemented wrong. 

- #1472 fixes that

File: packages/contracts-core/contracts/manager/BondingManager.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (trajan0x) [reply packages/contracts-core/contracts/manager/BondingManager.sol:91]: Awesome, let's address there if at all.

File: packages/contracts-core/contracts/libs/memory/Snapshot.sol [hunk 6]
```diff
@@ -161,13 +161,13 @@ library SnapshotLib {
...
     /// @param stateIndex    Index of Origin State in the Snapshot
-    function proofSnapRoot(bytes32 originRoot, uint32 domain, bytes32[] memory snapProof, uint256 stateIndex)
+    function proofSnapRoot(bytes32 originRoot, uint32 domain, bytes32[] memory snapProof, uint8 stateIndex)
         internal
         pure
         returns (bytes32)
     {
         // Index of "leftLeaf" is twice the state position in the snapshot
-        uint256 leftLeafIndex = stateIndex << 1;
+        uint256 leftLeafIndex = uint256(stateIndex) << 1;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/libs/memory/Snapshot.sol:173]: This was fixed in [7071fc7](https://github.com/synapsecns/sanguine/pull/1457/commits/7071fc785856832218b3b8ed19079902ee989d32)

File: packages/contracts-core/contracts/libs/memory/Snapshot.sol [hunk 7]
```diff
@@ -161,13 +161,13 @@
...
         // Check that "leftLeaf" index fits into Snapshot Merkle Tree
         if (leftLeafIndex >= (1 << SNAPSHOT_TREE_HEIGHT)) revert IndexOutOfRange();
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/libs/memory/Snapshot.sol:175]: This was fixed in [7071fc7](https://github.com/synapsecns/sanguine/pull/1457/commits/7071fc785856832218b3b8ed19079902ee989d32)

File: packages/contracts-core/test/utils/libs/SynapseStructs.t.sol [hunk 8]
```diff
@@ -227,6 +227,11 @@ library CastLib {
         request = RequestLib.encodeRequest({gasDrop_: rr.gasDrop, gasLimit_: rr.gasLimit, version_: rr.version});
     }
 
+    function boundRequest(RawRequest memory rr, uint96 maxGasDrop, uint64 maxGasLimit) internal pure {
+        rr.gasDrop = rr.gasDrop % maxGasDrop;
+        rr.gasLimit = rr.gasLimit % maxGasLimit;
```
[the remark this answers is not in the document]
Comment (ChiTimesChi) [reply packages/contracts-core/test/utils/libs/SynapseStructs.t.sol]: This was fixed in [17b3169](https://github.com/synapsecns/sanguine/pull/1457/commits/17b316985d6ff957a52dd47b703d3bf8e09d856e)

File: packages/contracts-core/contracts/manager/AgentManager.sol [hunk 9]
```diff
@@ -98,9 +98,12 @@ abstract contract AgentManager is MessagingBase, AgentManagerEvents, IAgentManag
         if (_agentDispute[notaryIndex].flag != DisputeFlag.None) revert NotaryInDispute();
         _disputes.push(OpenedDispute(guardIndex, notaryIndex, 0));
         // Dispute is stored at length - 1, but we store the index + 1 to distinguish from "not in dispute".
-        uint256 disputePtr = _disputes.length;
-        _agentDispute[guardIndex] = AgentDispute(DisputeFlag.Pending, uint88(disputePtr), address(0));
-        _agentDispute[notaryIndex] = AgentDispute(DisputeFlag.Pending, uint88(disputePtr), address(0));
+        // TODO: check if we really need to use 88 bits for dispute indexes. Every dispute ends up in one of
+        // the agents being slashed, so the number of disputes is limited by the number of agents (currently 2**32).
+        // Thus we can do the unsafe cast to uint88.
+        uint88 disputePtr = uint88(_disputes.length);
```
Comment (mikeyrf) [inline packages/contracts-core/contracts/manager/AgentManager.sol:104]: does this comment logic apply if at any point number of agents is changed (is this possible)?

**Review by mikeyrf — APPROVED** [review]
Overall looks good. May want to add some tests of edge cases where safe cast would cause reverts outside of the lib.

File: packages/contracts-core/contracts/manager/AgentManager.sol (hunk 9 above)
Comment (ChiTimesChi) [reply packages/contracts-core/contracts/manager/AgentManager.sol:104]: The overall logic stands: the amount of disputes is always limited by the max total amount of agents (every dispute removes an agent).

If the Agent Merkle tree is expanded, and more than 2**32 historical agents could exist, it's very likely that we would still go for something lower than 88 bits there, as adding of the new agents would be naturally limited by providing the bonds.

**Comment by ChiTimesChi** [comment]
Agreed on the edge cases, though some of them are not trivial to implement (like adding 2**32 entries to the array). Will add the lower hanging fruits though.
